Spark upsert table backfill support

pengding-stripe opened the pull request:
`bugfix`

### Summary
- `SegmentNameGenerator` for `UPLOADED_REALTIME` should get parameters from job spec instead of table config. This PR fixes this issue https://github.com/apache/pinot/issues/14083
- `getTableConfig` doesn't handle realtime table which looks like `{"REALTIME": {...)}`, `SparkSegmentGenerationJobRunner` uses it to decode table config and create segments

This PR fills gaps of backfilling upsert table with batch data using Spark

### Test
I uploaded a segment with one row of fake data using Spark ingestion pipeline, verified query returns fake data for this primary key `charge_id` now.
<img width="1367" alt="Screenshot 2024-11-13 at 1 56 36 PM" src="https://github.com/user-attachments/assets/af65f648-eaa9-4037-a9d0-5af6f0d2ea6f">

Jackie-Jiang commented on pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-common/src/main/java/org/apache/pinot/plugin/ingestion/batch/common/SegmentGenerationTaskRunner.java:
```diff
@@ -167,19 +171,25 @@ private SegmentNameGenerator getSegmentNameGenerator(SegmentGeneratorConfig segm
-        Preconditions.checkState(segmentGeneratorConfig.getUploadedSegmentPartitionId() != -1,
+        Preconditions.checkState(segmentNameGeneratorConfigs.get(SEGMENT_PARTITION_ID) != null,
             "Valid partition id must be set for uploaded realtime segment name generator");
         long creationTime;
```
Suggest renaming this as it is upload time instead of creation time.

pengding-stripe replied:
Sure, but I'm curious if this even matters? I think only partition id and table name actually matters for upsert table

Jackie-Jiang replied:
It matters to break tie when 2 segments contain the records with same timestamp (comparison value)

pengding-stripe replied:
but for upsert table same comparison value means same row, query should return same result

Jackie-Jiang reviewed (commented):
@rohityadav1993 Can you help take a look at this change? I saw you are using `SEGMENT_UPLOAD_TIME` instead of creation time in `BatchConfig` (introduced in #13837). Wondering why is it upload time instead of creation time? Also, since that field comes from the table config, do you expect user to pass in a different table config for each flink job?

pengding-stripe commented on pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-common/src/main/java/org/apache/pinot/plugin/ingestion/batch/common/SegmentGenerationTaskRunner.java:
```diff
@@ -62,6 +62,10 @@ public class SegmentGenerationTaskRunner implements Serializable {
   public static final String SEGMENT_NAME_TEMPLATE = "segment.name.template";
 
+  // For UploadedRealtimeSegmentNameGenerator
+  public static final String SEGMENT_PARTITION_ID = "segment.partitionId";
+  public static final String SEGMENT_UPLOAD_TIME_MS = "segment.uploadTimeMs";
```
yeah I see everything else is duplicated here so I also duplicate this one, I can update to use BatchConfigProperties

rohityadav1993 replied:
`SegmentGeneratorConfig.segmentCreationTime` can be reused as `SEGMENT_UPLOAD_TIME`. Every run of ingestion job should have a time associated with it and should populate it.

Since Flink can also be run in batch mode, e.g. hourly/daily; start/end time can be used as `creationTime`. This provides deterministic segment names and one run's segments will have same timestamp in segment name.

Flink uses table config but also overrides transient properties like partitionId, seqId, creationTime in [FlinkSegmentWriter](https://github.com/apache/pinot/pull/13837/files#diff-72b89e65ea292a808b5d8a902d8a64adb87834534414d5b9d0e46c4725a0455cL126) through flink job config `SEGMENT_UPLOAD_TIME_MS`. This way fixed properties(like prefix) can be part of table config and transient properties should come from the ingestion job.

pengding-stripe replied:
To use `SegmentGeneratorConfig.segmentCreationTime` you have to call `SegmentGeneratorConfig.setCreationTime()` here first. Are you suggesting calling it here with system time instead of letting user customize it? I don't think we can use `SegmentGeneratorConfig.segmentCreationTime` directly without setting it. 
Spark is quite similar, it can also run in batch mode and one run's segments will have same timestamp in segment name

rohityadav1993 commented on pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-common/src/main/java/org/apache/pinot/plugin/ingestion/batch/common/SegmentGenerationTaskRunner.java:
```diff
@@ -167,19 +167,25 @@ private SegmentNameGenerator getSegmentNameGenerator(SegmentGeneratorConfig segm
-        Preconditions.checkState(segmentGeneratorConfig.getUploadedSegmentPartitionId() != -1,
+        Preconditions.checkState(segmentNameGeneratorConfigs.get(BatchConfigProperties.SEGMENT_UPLOAD_TIME_MS) != null,
+            "Upload time must be set for uploaded realtime segment name generator");
+        Preconditions.checkState(segmentNameGeneratorConfigs.get(BatchConfigProperties.SEGMENT_PARTITION_ID) != null,
```
How will the partitionId be provided by the task? The partitionId means the data is partitioned and same as stream that is ingesting to realtime table.

There should be a corresponding code in spark to partition data and derive id accordingly.

pengding-stripe replied:
Users need to partition data themselves. We have a spark job to replicate how stream ingestion partition data and put partitioned data in a path with partition id. e.g. `s3://.../partition=0/partitioned-data.parquet`
Then create segment task can get partition id from the path. So there will be one spark task per partition.

rohityadav1993 commented on pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-common/src/main/java/org/apache/pinot/plugin/ingestion/batch/common/SegmentGenerationTaskRunner.java:
```diff
@@ -167,19 +167,25 @@ private SegmentNameGenerator getSegmentNameGenerator(SegmentGeneratorConfig segm
         return new InputFileSegmentNameGenerator(segmentNameGeneratorConfigs.get(FILE_PATH_PATTERN),
             segmentNameGeneratorConfigs.get(SEGMENT_NAME_TEMPLATE), inputFileUri, appendUUIDToSegmentName);
       case BatchConfigProperties.SegmentNameGeneratorType.UPLOADED_REALTIME:
-        Preconditions.checkState(segmentGeneratorConfig.getCreationTime() != null,
```
uploadTime is a property of the ingestionJob to provide deterministic names to segments based on when the job ran and re-running replaces the segments. Moving this to segmentNameGeneratorConfig means it is no more a job property and can be set independently.

The upstream job should have the responsibility to either use its creation time or default to currentTime/any other time.

pengding-stripe replied:
If you look at the original issue: https://github.com/apache/pinot/issues/14083
The reporter got an error `Creation time must be set for uploaded realtime segment name generator`. I think this means it's not a default property of the ingestion job, otherwise he shouldn't get this error. So I moved it to segmentNameGeneratorConfig and simply use current system time in the upstream job

rohityadav1993 replied:
Can we use `segmentNameGeneratorConfigs.get(BatchConfigProperties.SEGMENT_UPLOAD_TIME_MS)` and also fall back to `segmentGeneratorConfig.getCreationTime() for ` [UploadedRealtimeSegmentName](https://github.com/apache/pinot/blob/07962b0adefe2af8c5698d8d674f35a395d7b129/pinot-common/src/main/java/org/apache/pinot/common/utils/UploadedRealtimeSegmentName.java#L42)

I feel this multiple places for providing create time can be confusing.

pengding-stripe commented:
@Jackie-Jiang there is an integration test failure but I don't think that's related to my change, how can I retrigger that test?

Jackie-Jiang commented:
@pengding-stripe No worry, the failure is not related to this PR

pengding-stripe commented:
@rohityadav1993 I did more tests to push multiple segments, you can see that these segments are for different partitions with same timestamp. Because I submitted multiple job specs to a spark app and ran in parallel. This PR doesn't cover segment partition part though, it just fixes some bugs for Spark upsert table backfill. Users need to have their own upstream jobs to partition data correctly.
<img width="733" alt="Screenshot 2024-11-15 at 10 05 17 AM" src="https://github.com/user-attachments/assets/1f400e04-67f6-4971-ba4c-1dde5682d817">

rohityadav1993 reviewed (approved):
Thanks for the tests @pengding-stripe, lgtm. Will tag @klsince for approval as he also has context for [UploadedRealtimeSegmentName](https://github.com/apache/pinot/blob/07962b0adefe2af8c5698d8d674f35a395d7b129/pinot-common/src/main/java/org/apache/pinot/common/utils/UploadedRealtimeSegmentName.java#L42).

pengding-stripe commented:
@rohityadav1993 I already updated to use `segmentNameGeneratorConfigs` in previous commit, also updated to fall back to `segmentGeneratorConfig.getCreationTime()`, thanks for the review
